### PR TITLE
storage: categorize more ssh errors correctly

### PR DESCRIPTION
**Pull request by guswynn**
Closes: https://github.com/MaterializeInc/materialize/issues/22280

- ensures all kafka sources (including ones that use csr) and sinks categorize errors correctly
- tests ssh error categorization for sinks

### Motivation
  * This PR adds a known-desirable feature.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by jkosh44 — COMMENTED** [review]
Adapter bits LGTM

File: src/storage-client/src/sink.rs [hunk 1]
```diff
@@ -336,17 +344,16 @@ pub async fn build_kafka(
...
+        KafkaConsistencyConfig::Progress { topic } => ensure_kafka_topic(
+            &client,
+            topic,
+            1,
+            connection.replication_factor,
+            KafkaSinkConnectionRetention::default(),
+        )
+        .await
+        .check_ssh_status(client.inner().context())
+        .add_context("error registering kafka consistency topic for sink")?,
```
Comment (benesch) [inline src/storage-client/src/sink.rs:356]: This doesn't look terribly unmanageable, but it does seem easy for someone for someone to refactor and lose a `check_ssh_status` call, or add a new call to a Kafka client method and forget to call `check_ssh_status`. I don't have anything better to propose, though.

File: src/storage-client/src/sink.rs [hunk 2]
```diff
@@ -283,18 +286,22 @@ pub async fn build_kafka(
...
         let latest_ts = determine_latest_progress_record(
             format!("build_kafka_{}", sink_id),
             progress_topic.name().to_string(),
             ProgressKey::new(sink_id),
-            Arc::new(progress_client),
+            Arc::clone(&progress_client),
         )
-        .await?;
+        .await
+        .check_ssh_status(progress_client.client().context())?;
```
Comment (benesch) [inline src/storage-client/src/sink.rs:297]: Is it always valid to override the error with the SSH error? I guess so? Like, if the SSH tunnel is broken, you're screwed, even if there was another error..

**Review by benesch — COMMENTED** [review]
Deferring the rest of the review to @petrosagg! Thanks for iterating on the approach!

File: src/storage-client/src/sink.rs (hunk 2 above)
Comment (guswynn) [reply src/storage-client/src/sink.rs:297]: Yeah, thats the idea!

File: src/storage-client/src/sink.rs (hunk 1 above)
Comment (guswynn) [reply src/storage-client/src/sink.rs:356]: Yeah...i couldn't think of anything better...

**Review by petrosagg — APPROVED** [review]

File: src/sql/src/pure/error.rs [hunk 3]
```diff
@@ -233,8 +233,12 @@ impl KafkaSinkPurificationError {
 pub enum CsrPurificationError {
     #[error("{0} is not a CONFLUENT SCHEMA REGISTRY CONNECTION")]
     NotCsrConnection(FullItemName),
+    // `mz_storage_types::Errors::CsrConnectError` has an `anyhow::Error` that is not cloneable.
     #[error("client errored")]
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/sql/src/pure/error.rs:237]: ill check if its possible easily

[the remark this answers is not in the document]
Comment (guswynn) [reply src/sql/src/pure/error.rs:237]: this wasnt too bad!

**Pull request merged**